harness/determined/layers/_workload_manager.py

```diff
@@ -254,35 +260,41 @@ def yield_checkpoint_model(
         start_time = _current_timestamp()
 
         # Only the chief container should checkpoint.
-        if self.rendezvous_info.get_rank() == 0:
-            with self.storage_mgr.store_path() as (storage_id, path):
-
-                def _respond(checkpoint_info: workload.Response) -> None:
-                    checkpoint_info = cast(Dict[str, Any], checkpoint_info)
-                    metadata = storage.StorageMetadata(
-                        storage_id,
-                        storage.StorageManager._list_directory(path),
-                        checkpoint_info.get("framework", ""),
-                        checkpoint_info.get("format", ""),
-                    )
+        if self.rendezvous_info.get_rank() != 0:
+            respond(workload.Skipped())
+            return
 
-                    logging.info("Saved trial to checkpoint {}".format(metadata.storage_id))
-                    self.tensorboard_mgr.sync()
+        # Save the workload completed message for after checkpoint upload completes.
+        message = None  # type: Optional[workload.Response]
 
-                    message: workload.Response = {
-                        "type": "WORKLOAD_COMPLETED",
-                        "workload": wkld,
-                        "start_time": start_time,
-                        "end_time": _current_timestamp(),
-                        "metrics": metadata,
-                    }
+        def _respond(checkpoint_info: workload.Response) -> None:
+            checkpoint_info = cast(Dict[str, Any], checkpoint_info)
+            metadata = storage.StorageMetadata(
+                storage_id,
+                storage.StorageManager._list_directory(path),
+                checkpoint_info.get("framework", ""),
+                checkpoint_info.get("format", ""),
+            )
 
-                    respond(message)
+            logging.info("Saved trial to checkpoint {}".format(metadata.storage_id))
+            self.tensorboard_mgr.sync()
 
-                yield wkld, [pathlib.Path(path)], _respond
+            nonlocal message
+            message = {
+                "type": "WORKLOAD_COMPLETED",
+                "workload": wkld,
+                "start_time": start_time,
+                "end_time": _current_timestamp(),
+                "metrics": metadata,
+            }
 
-        else:
-            respond(workload.Skipped())
+        with self.storage_mgr.store_path() as (storage_id, path):
+            yield wkld, [pathlib.Path(path)], _respond
+
+        check_not_none(message, "response function did not get called")
```

Comment:
maybe drop a comment that this necessarily works (and outs a programming error) because `_respond` must be called before `next()` is called on the parent generator, so `message` _should_ exist.

Reply:
done